artifact/gem5art/artifact/artifact.py

```diff
@@ -75,103 +106,139 @@ class Artifact:
-        self['type'] = typ
-        self['documentation'] = cleandoc(documentation)
-        if len(self['documentation']) < 10: # 10 characters is arbitrary
+        data['name'] = name
+        data['type'] = typ
+        data['documentation'] = cleandoc(documentation)
+        if len(data['documentation']) < 10: # 10 characters is arbitrary
             raise Exception(cleandoc("""Must provide longer documentation!
-                This documentation is how your future self will remember what
+                This documentation is how your future data will remember what
                 this artifact is and how it was created."""))
 
-        self['command'] = cleandoc(command)
+        data['command'] = cleandoc(command)
+
+        data['time'] = time.time()
 
-        self['path'] = path
+        data['path'] = path
         if os.path.isfile(path):
-            self['hash'] = getHash(path)
-            self['git'] = None
+            data['hash'] = getHash(path)
+            data['git'] = {}
         elif os.path.isdir(path):
-            self['git'] = getGit(path)
-            self['hash'] = self['git']['hash']
+            data['git'] = getGit(path)
+            data['hash'] = data['git']['hash']
         else:
             raise Exception("Path {} doesn't exist".format(path))
 
-        self['cwd'] = cwd
+        data['cwd'] = cwd
         if not os.path.exists(cwd):
             raise Exception("cwd {} doesn't exits.".format(cwd))
 
-        self['inputs'] = [i._id for i in inputs]
+        data['inputs'] = [i._id for i in inputs]
 
-        if self['hash'] in _db:
-            old_artifact = Artifact(_db.get(self['hash']))
-            self['_id'] = old_artifact._id
+        if data['hash'] in _db:
+            old_artifact = Artifact(_db.get(data['hash']))
+            data['_id'] = old_artifact._id
 
             # Now that we have a complete object, construct it
-            self = cls(self)
+            self = cls(data)
             if old_artifact != self:
                 print(f"Current: {vars(self)}")
                 print(f"From DB: {vars(old_artifact)}")
                 raise Exception("Found matching hash in DB, but object "
                     "doesn't match. Use the UUID constructor instead.")
         else:
-            self['_id'] = uuid4()
+            data['_id'] = uuid4()
 
             # Now that we have a complete object, construct it
-            self = cls(self)
-            _db.put(self._id, self)
+            self = cls(data)
+            _db.put(self._id, self._getSerializable())
 
             # Upload the file if there is one.
             if os.path.isfile(self.path):
                 _db.upload(self._id, self.path)
 
         return self
 
-    def __init__(self, other):
+    def __init__(self, other: Union[str, UUID, Dict[str, Any]]) -> None:
         """Constructs the object from the database based on a UUID or
         dictionary from the database
         """
+        if isinstance(other, str):
```

Comment:
I guess we need some way to assert (or to document) that "if other is a string then it must be a UUID"

Reply:
If I add "if other is a string it must be constructable as a UUID" to the documentation would that be enough?